### PR TITLE
Gg 29423

### PR DESCRIPTION
**Pull request by glukos**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java [hunk 1]
```diff
@@ -1020,23 +1020,18 @@ private Metas getOrAllocateCacheMetas() throws IgniteCheckedException {
...
-        FileWALPointer minPtr = null;
+        Map<Integer, Long> prtsCounter = new HashMap<>();
 
         for (int i = 0; i < partCntrs.size(); i++) {
             int p = partCntrs.partitionAt(i);
             long initCntr = partCntrs.initialUpdateCounterAt(i);
 
-            FileWALPointer startPtr = (FileWALPointer)database.checkpointHistory().searchPartitionCounter(
-                grp.groupId(), p, initCntr);
+            prtsCounter.put(p, initCntr);
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java:1029]: shouldn't it be partsCounters?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 2]
```diff
@@ -90,6 +89,9 @@
     /** Maximum size of WAL archive (in bytes) */
     private final long maxWalArchiveSize;
 
+    /** Map stores the earliest checkpoint for each partition from particular group. */
+    private Map<T2<Integer, Integer>, CheckpointEntry> erliestCp = new HashMap<>();
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:93]: typo: earliestCp

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 3]
```diff
@@ -114,6 +116,15 @@ public CheckpointHistory(GridKernalContext ctx) {
     public void initialize(List<CheckpointEntry> checkpoints) {
         for (CheckpointEntry e : checkpoints)
             histMap.put(e.timestamp(), e);
+
+        for (Long timestamp : checkpoints(false)) {
+            try {
+                updateErliestCpMap(entry(timestamp));
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:122]: typo

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 4]
```diff
@@ -181,9 +192,66 @@ public WALPointer firstCheckpointPointer() {
...
+     *
+     * @param entry Checkpoint entry.
+     */
+    private void updateErliestCpMap(CheckpointEntry entry) {
+        try {
+            Map<Integer, CheckpointEntry.GroupState> states = entry.groupState(cctx);
+
+            HashMap<Integer, Boolean> applicableGrps = new HashMap<>();
+
+            Iterator<Map.Entry<T2<Integer, Integer>, CheckpointEntry>> iter = erliestCp.entrySet().iterator();
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:211]: We can use GroupPartitionId as a key

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 5]
```diff
@@ -359,21 +440,105 @@ private long absFileIdx(CheckpointEntry pointer) {
...
+                    if (ptr == null) {
+                        throw new IgniteCheckedException("Could not find start pointer for partition [part="
+                            + entry.getKey() + ", partCntrSince=" + entry.getValue() + "]");
+                    }
+
+                    if (minPtr == null || ptr.compareTo(minPtr) < 0)
+                        minPtr = ptr;
+                }
+
+            }
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:484]: Can we break; here if modifiedPartsCounter is already empty?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 6]
```diff
@@ -359,21 +440,105 @@ private long absFileIdx(CheckpointEntry pointer) {
...
+    @Nullable public Map<T2<Integer, Integer>, CheckpointEntry> searchCheckpointEntry(Map<T2<Integer, Integer>, Long> searchCntrMap) {
+        if (F.isEmpty(searchCntrMap))
+            return Collections.emptyMap();
 
-        if (entry == null)
-            return null;
+        Map<T2<Integer, Integer>, Long> modifiedSearchMap = new HashMap<>();
+        modifiedSearchMap.putAll(searchCntrMap);
+
+        Map<T2<Integer, Integer>, CheckpointEntry> res = new HashMap<>();
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:510]: GroupPartitionId

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java [hunk 7]
```diff
@@ -1812,7 +1815,7 @@ private boolean safeToUpdatePageMemories() {
 
         Map</*grpId*/Integer, Map</*partId*/Integer, /*updCntr*/Long>> grpPartsWithCnts = new HashMap<>();
 
-        for (Map.Entry<Integer, T2</*reason*/String, Map</*partId*/Integer, CheckpointEntry>>> e : earliestValidCheckpoints.entrySet()) {
+        for (Map.Entry<Integer, T2</*reason*/ReservationReason, Map</*partId*/Integer, CheckpointEntry>>> e : earliestValidCheckpoints.entrySet()) {
             int grpId = e.getKey();
 
             if (e.getValue().get2() == null) {
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1821]: Can we get rid of reservedForExchange map? We actually need one earliestPointer to perform release later.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 8]
```diff
@@ -181,9 +192,66 @@ public WALPointer firstCheckpointPointer() {
...
+
         histMap.put(entry.timestamp(), entry);
     }
 
+    /**
+     * Update map which stored the earliest checkpoint each partitions for groups.
+     *
+     * @param entry Checkpoint entry.
+     */
+    private void updateErliestCpMap(CheckpointEntry entry) {
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:205]: typo: earliest

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 9]
```diff
@@ -255,7 +323,20 @@ private boolean removeCheckpoint(CheckpointEntry checkpoint) {
...
+        CheckpointEntry deletedCpEntry = histMap.remove(checkpoint.timestamp());
+
+        CheckpointEntry oldestCpInHistory = firstCheckpoint();
+
+        Iterator<Map.Entry<T2<Integer, Integer>, CheckpointEntry>> iter = erliestCp.entrySet().iterator();
+
+        log.info("Remove cp " + deletedCpEntry.checkpointId());
+
+        while (iter.hasNext()) {
+            Map.Entry<T2<Integer, Integer>, CheckpointEntry> grpParCp = iter.next();
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:335]: grpPerCp ?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java (hunk 7 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1819]: I have done this comment.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java (hunk 1 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java]: Yes, renamed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 8 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 4 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 9 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java]: Renamed to grpPartPerCp

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 5 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:482]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 6 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 10]
```diff
@@ -90,6 +90,9 @@
     /** Maximum size of WAL archive (in bytes) */
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:90]: Lots of reason string constants in this class became unused, we can safely remove them.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 11]
```diff
@@ -181,9 +193,102 @@ public WALPointer firstCheckpointPointer() {
      * @param entry Entry to add.
      */
     public void addCheckpoint(CheckpointEntry entry) {
+        addCpToEarliestCpMap(entry);
+
         histMap.put(entry.timestamp(), entry);
     }
 
+    /**
+     * Update map which stored the earliest checkpoint each partitions for groups.
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:202]: Updates map which stores the earliest checkpoint for each partitions in every group.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 12]
```diff
@@ -90,6 +90,9 @@
     /** Maximum size of WAL archive (in bytes) */
     private final long maxWalArchiveSize;
 
+    /** Map stores the earliest checkpoint for each partition from particular group. */
+    private Map<GroupPartitionId, CheckpointEntry> earliestCp = new HashMap<>();
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:94]: Can be final

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 13]
```diff
@@ -181,9 +193,102 @@ public WALPointer firstCheckpointPointer() {
...
+     * @param grpId Group id.
+     * @return Checkpoint witch it'd be marked as inapplicable.
+     */
+    public CheckpointEntry lastCheckpointMarkingAsInapplicable(Integer grpId) {
+        synchronized (earliestCp) {
+            CheckpointEntry lastCp = lastCheckpoint();
+
+            Iterator<Map.Entry<GroupPartitionId, CheckpointEntry>> iter = earliestCp.entrySet().iterator();
+
+            while (iter.hasNext()) {
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:252]: Can be replaced with removeIf

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 14]
```diff
@@ -181,9 +193,102 @@ public WALPointer firstCheckpointPointer() {
...
+    private void updateEarliestCpMap(CheckpointEntry entry) {
+        try {
+            Map<Integer, CheckpointEntry.GroupState> states = entry.groupState(cctx);
+
+            Iterator<Map.Entry<GroupPartitionId, CheckpointEntry>> iter = earliestCp.entrySet().iterator();
+
+            while (iter.hasNext()) {
+                Map.Entry<GroupPartitionId, CheckpointEntry> grpPartCp = iter.next();
+
+                Integer grpId = grpPartCp.getKey().getGroupId();
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:215]: Why not int?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 15]
```diff
@@ -181,9 +193,102 @@ public WALPointer firstCheckpointPointer() {
...
+
+                if (!isCheckpointApplicableForGroup(grpId, entry)) {
+                    iter.remove();
+
+                    continue;
+                }
+
+                Integer part = grpPartCp.getKey().getPartitionId();
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:223]: Why not int?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 16]
```diff
@@ -407,125 +608,58 @@ private long absFileIdx(CheckpointEntry pointer) {
      *
      * @param groupsAndPartitions Groups and partitions to find and reserve earliest valid checkpoint.
      *
-     * @return Map (groupId, Map (partitionId, earliest valid checkpoint to history search)).
+     * @return Map (groupId, Reason (the reason why reservation cannot be make deeper): Map
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:611]: cannot be made deeper

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 15 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:223]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 16 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:611]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 14 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:215]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 12 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:94]: Made it final.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 11 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:202]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 10 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:90]: Removed all *_REASON constants.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java (hunk 13 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:252]: Done

**Pull request closed without merging**